apps/google-analytics-4/lambda/src/middlewares/service-account-key-provider.ts

```diff
@@ -0,0 +1,84 @@
+import { RequestHandler } from 'express';
+import { ServiceAccountKeyFile, ServiceAccountKeyId } from '../types';
+
+export class MissingServiceAccountKeyHeader extends Error {}
+export class InvalidServiceAccountKey extends Error {}
+
+export const serviceAccountKeyProvider: RequestHandler = (req, _res, next) => {
+  const serviceAccountKeyIdHeaderValue = req.headers['x-contentful-serviceaccountkeyid'];
+  if (!serviceAccountKeyIdHeaderValue || typeof serviceAccountKeyIdHeaderValue !== 'string') {
+    throw new MissingServiceAccountKeyHeader(
+      "missing or incorrectly formatted header 'x-contentful-serviceaccountkeyid'"
+    );
+  }
+  const serviceAccountKeyId = decodeServiceAccountHeader(serviceAccountKeyIdHeaderValue);
+  assertServiceAccountKeyId(serviceAccountKeyId);
+  req.serviceAccountKeyId = serviceAccountKeyId;
+
+  // Pulling the account key from the header will be removed later and replaced with fetching
+  // the code directly from secret storage!
+  const serviceAccountKeyHeaderValue = req.headers['x-contentful-serviceaccountkey'];
+  if (!serviceAccountKeyHeaderValue || typeof serviceAccountKeyHeaderValue !== 'string') {
+    throw new MissingServiceAccountKeyHeader(
+      "missing or incorrectly formatted header 'x-contentful-serviceaccountkey'"
+    );
+  }
+  const serviceAccountKey = decodeServiceAccountHeader(serviceAccountKeyHeaderValue);
+  assertServiceAccountKeyFile(serviceAccountKey);
+  req.serviceAccountKey = serviceAccountKey;
+
+  next();
+};
+
+const decodeServiceAccountHeader = (value: string) => {
+  const asciiStr = Buffer.from(value, 'base64').toString('ascii');
+  try {
+    return JSON.parse(asciiStr);
+  } catch (e) {
+    throw new InvalidServiceAccountKey('Improperly formatted service account key', { cause: e });
+  }
+};
+
+function assertServiceAccountKeyId(value: unknown): asserts value is ServiceAccountKeyId {
+  const requiredFields = ['clientEmail', 'projectId', 'id'];
+  for (const requiredField of requiredFields) {
+    if (!value || typeof value !== 'object' || !(requiredField in value)) {
+      throw new InvalidServiceAccountKey(
+        `'${requiredField}' is missing from Service Account Key header`
```

Comment:
nit: remove ` header` here so this still makes sense once this is being extracted from secrets manager:

```
`'${requiredField}' is missing in Service Account Key data`
```

Reply:
Also, this same feedback in the other instances of this messaging below, too

Reply:
Sure, can do. (FWIW I believe the secret storage bit will be entirely new / non backward compatible error messages, which i consider fine for this development stage 😉 )